zmap/zgrab2
# Implements postgres zgrab2 module

## What it does ##

* Provides a zgrab2 module for postgres
* Integration tests for postgres zgrab module
* Various fixes / refactoring

## How to test it ##

Travis should pick it up, but you can set up a basic postgres docker image (no SSL) by running
 * `docker run -itd -p 5432:5432 --rm --name zgrab_pg -e POSTGRES_PASSWORD postgres:10.1`

Then, you can run the scanner against localhost:
* `echo "127.0.0.1" | cmd/zgrab2/zgrab2 postgres --timeout 10`

You should see some server-generated strings in the output, e.g. `"FATAL:  unsupported frontend protocol 0.0: server supports 2.0 to 3.0"`

Or, as always, you can test the integration tests by running `pushd integration_tests/postgres && ./setup.sh && ./test.sh && ./cleanup.sh && popd`, or to run them alongside the rest of the tests, `pushd integration_tests &&./setup.sh && ./test.sh && ./cleanup.sh && popd`

## Notes ##

I'll try to call anything important out in comments to the code, but the main changes are:

 * **integration_tests/**:
   * *Cleanup*: Standardize container/image names (e.g. testmysql, sshtest -> zgrab_mysql, zgrab_ssh)
   * *Cleanup*: Consistently use /usr/bin/env bash rather than /bin/bash
   * *Bugfix*: Ensure set -e gets set everywhere it is needed  (this was leading to false test-passes)
   * *Bugfix*: `test.sh`: Catch / log all validation errors
   * *Bugfix*: `test.sh`: Catch schematically-valid but non-successful scans (e.g. status = connection-refused)
   * *New implementation*: `new.sh`: add schema skeleton

 * **postgres/**:
   * *New implementation*: Initial version of scanner
   * *New implementation*: integration tests

* **schemas/**:
  * *New implementation*: `zcrypto.py`: fill out `client_hello`

* **misc**
   * *Refactor*: Move ScanStatus & related to status.go (previously scattered around processing.go, module.go, utility.go)
   * *Refactor*: Add `ScanTarget.Open(*BaseFlags)` method (so every module doesn't need to re-write the same `Dial("${ScanTarget.IP}:${Config.BaseFlags.Port}")` and `SetReadDeadline` / `SetWriteDeadline` code (and, so scanners no longer need to be aware of the host/port/timeout)
   * *Bugfix*: `tls.go`: Better implementation of TLSConnection.GetLog() -- keep the pointer around and populate it as we go
